### DISCUSSION TITLE
Just a question, not an issue

### DISCUSSION THREAD

**Discussion by BradenDKelly** [ClapeyronThermo/Clapeyron.jl#142] [Q&A]
I am just curious what the roadmap is for flash calculations? Will you be implementing such things as the Tangent Plane Criteria etc and essentially solve for the number of possible phases automatically? Do you need or want help?

**Comment by pw0908** [comment]
Hi Braden,

It's not currently clear how we're going to implement flash algorithms. Andrés has already started with some preliminary work and I've created a (barely) functional one on my electrolytemodels branch.

We initially wanted to code the HELD algorithm since this method does not require any initial guesses or specified number of phases. However, understanding HELD is proving to be its own major undertaking.

We would love to have some help with this if you would be willing to provide any! I'd imagine you're more familiar with Michelsen's algorithms which seem to be the standard method used in most packages?

Best,
Pierre

**Comment by BradenDKelly** [comment]
Hi Pierre,

It would be a stretch to say I am familiar with Michelsens, but I do have Matlab code somewhere from many years ago that I made after taking his course. I am not sure how well I trust it, my knowledge of unit tests at that time was non existent. I can take a look at it and see what I can come up with on a forked branch over some time. I would love to be able to use the SAFT's for the work I am doing, and flash calculations are always useful, particularly for determining number of phases without assuming. I have bad luck with assumptions.

Cheers,
Braden

**Comment by pw0908** [comment]
Hi Braden,

That's more than any of us! If you can translate your code to Julia, we can refine it over time to get at least a working implementation.

I agree that the flash algorithm is something that's very useful in any thermo modelling software and it's been a gap in our packages we've been meaning to fill for a while...

It's amazing that you had to code it for a course! I'm guessing this was at the DTU? 

Best regards,

Pierre

**Comment by BradenDKelly** [comment]
Hi Pierre,

It was back in 2014 at DTU. I am not sure if they still do it, but it was a once a year, 2 week PhD course (I was in my Masters then). I flew in for the two weeks. It was very good. Theory in the morning, coding in the afternoon. We covered a good chunk of Michelsen and Mollerup's textbook. There were some TA's to help as well. There were a few long nights, particularly by the end once complexity was at a maximum. We also got a week at home afterwards to finish things off and submit the assignments. It was without a doubt the best 2 week course I have taken. It was very involved and I left feeling like I was significantly better off than when I arrived.

RIP Michelsen <3

Cheers,

Braden

**Comment by longemen3000** [comment]
Hi,

the preliminary work done at the moment is a K-values rachford rice solver, based on Gaganis, V. (2021) "Solution of the Rachford Rice equation using perturbation analysis". it works and is tested at the moment, but im looking on what to return on one phase zones, and i'm willing to change those results if needed (mainly to support negative flash)

it is ok if we leave this issue open to keep track on any advances on the TP flash algorithm?

Best,

Andrés

**Comment by moyner** [comment]
Hi,

Thanks for the great work on this package! I just recently put out a [lightweight MIT licensed package](https://github.com/moyner/MultiComponentFlash.jl) for two-phase (p-T) flash of mixtures.

The package is tailored more towards coupling to flow simulations than Clapeyron.jl (so at the moment only cubics, but there are options for zero-allocating flash and everything is type stable). It contains flash with Newton/SSI, Rachford-Rice and stability testing. I thought it might be of interest if you are working on flash for more complex SAFT-like systems and there might be some opportunities for interop.

Cheers,
Olav
